lib/pf/UnifiedApi/Controller/Config/Roles.pm

```diff
@@ -26,7 +25,239 @@ has 'primary_key' => 'role_id';
+    node,
+    (SELECT category_id from node_category WHERE name = ?) as old_nc,
+    (SELECT category_id from node_category WHERE name = ?) as new_nc
+SET node.bypass_role_id = new_nc.category_id
+WHERE
+    old_nc.category_id IS NOT NULL AND
+    new_nc.category_id IS NOT NULL AND
+    node.bypass_role_id = old_nc.category_id;
+SQL
+
+my $REASSIGN_PASSWORD_CATEGORY = <<SQL;
+UPDATE
+    password,
+    (SELECT category_id from node_category WHERE name = ?) as old_nc,
+    (SELECT category_id from node_category WHERE name = ?) as new_nc
+SET password.category = new_nc.category_id
+WHERE
+    old_nc.category_id IS NOT NULL AND
+    new_nc.category_id IS NOT NULL AND
+    password.category = old_nc.category_id;
+SQL
+
+sub reassign {
+    my ($self) = @_;
+    my ($error, $data) = $self->get_json;
+    if (defined $error) {
+        return $self->render_error(400, "Bad Request : $error");
+    }
+
+    my @errors;
+    my $old = $self->id;
+    my $new = $data->{id};
+    $self->check_reassign_args($old, $new, \@errors);
+    if (@errors) {
+        return $self->render_error(422, "Unable to reassign role", \@errors);
+    }
+
+    $self->reassign_role_with_sql(\@errors, "pf::dal::node", $REASSIGN_NODE_CATEGORY_ID, $old, $new, "node category");
+    $self->reassign_role_with_sql(\@errors, "pf::dal::node", $REASSIGN_NODE_BYPASS_ROLE_ID, $old, $new, "node bypass role");
+    $self->reassign_role_with_sql(\@errors, "pf::dal::person", $REASSIGN_PASSWORD_CATEGORY, $old, $new, "password category");
+    $self->reassign_role_config_store(\@errors, "pf::ConfigStore::AdminRoles", $old, $new, qw(allowed_roles allowed_node_roles));
+    $self->reassign_role_config_store(\@errors, "pf::ConfigStore::Scan", $old, $new, qw(categories));
+    $self->reassign_role_config_store(\@errors, "pf::ConfigStore::Provisioning", $old, $new, qw(category role_to_apply));
+    $self->reassign_role_config_store(\@errors, "pf::ConfigStore::SelfService", $old, $new, qw(roles_allowed_to_unregister device_registration_roles));
+    $self->reassign_role_config_store(\@errors, "pf::ConfigStore::BillingTiers", $old, $new, qw(roles_allowed_to_unregister device_registration_roles));
+    $self->reassign_role_config_store(\@errors, "pf::ConfigStore::Firewall_SSO", $old, $new, qw(categories));
+    $self->reassign_role_config_store_switch(\@errors, $old, $new);
+    if (@errors) {
+        return $self->render_error(422, "Unable to reassign role", \@errors);
+    }
+
+    return $self->render(200, json => {});
+}
+
+sub reassign_role_config_store {
+    my ($self, $errors, $class, $old, $new, @fields) = @_;
+    my $cs = $class->new;
+    my $i = 0;
+    my $cachedConfig = $cs->cachedConfig;
+    for my $sect ($cs->_Sections()) {
+        for my $f (@fields) {
+            next if !$cachedConfig->exists($sect, $f);
+            my $values = $cachedConfig->val($sect, $f);
+            my @roles = split(/\s*,\s*/, $values);
+            my @new_roles = map { $_ eq $old ? $new : $_ } @roles;
```

Comment:
We should make this list unique to prevent the role from being there twice if it was already part of the list

Reply:
Ok